File: clients/python/clausters/defs/expr.py
"""What "an expression" is, as a type.

An **expression** is something that composes a DSP graph rather than a value:
a `clausters.defs.ugens.Ugen` graph, a channel list, a Faust
`clausters.defs.signals.Signal` or a `clausters.defs.boxes.Box`. It is what
the ambient verbs accept where a def is wanted, and what
`clausters.defs.asdef.as_def` coerces.

The distinction the base above this one cannot make: `clausters.base.absobject.
AbstractObject` is the *operator* protocol, and the value side
(`clausters.base.builtins`) shares it — the same written expression composes
concrete numbers or a graph depending on the subclass. `Expr` is the half that
composes a graph.

Its two branches are the two def families, which are peers: `SynthExpr` for the
UGen graph (`Ugen`, `Control`, `ChannelList`) and `FaustExpr` for Faust
(`Signal`, `Box`). A single engine-neutral base would make one family "the
graph" and the other the exception; `Signal` and `Box` do not compose with each
other either, so their shared roof buys a real dispatch. The name avoids
``Graph*``: `clausters.defs.graphdef.GraphDef` is the third def family, so in
this package "graph" already means a configuration of member nodes wired by
buses.

Not every composable thing is one: `clausters.defs.pv_expr.PvExpr` composes a
symbolic **per-bin** expression that serializes to the token list ``PV_Kernel``
interprets. It is never a graph node — it cannot be a def root nor a UGen input
— so it stays outside this hierarchy and the verbs keep rejecting it.

These are markers: they carry no behavior, and the four composition hooks stay
in the classes that implement them.
"""

from ..base.absobject import AbstractObject


class Expr(AbstractObject):
    """Something that composes a DSP graph rather than a value — a `SynthExpr`
    or a `FaustExpr`. The type the ambient verbs (`clausters.play`,
    `clausters.plot`, `clausters.render`) accept as a bare expression."""


class SynthExpr(Expr):
    """An expression of the **UGen graph**: a `clausters.defs.ugens.Ugen`, a
    `clausters.defs.ugens.Control`, or a `clausters.defs.ugens.ChannelList` of
    them. What `clausters.defs.synthdef.SynthDef` serializes and what
    ``out``/``replace_out``/``out_ctl`` accept."""


class FaustExpr(Expr):
    """An expression of a **Faust graph**: a `clausters.defs.signals.Signal`
    (signal API) or a `clausters.defs.boxes.Box` (box API). What
    `clausters.defs.faustdef.FaustDef` compiles."""

File: clients/web/src/defs/expr.ts
// What "a Faust expression" is, as a type (mirrors `clausters/defs/expr.py`).
//
// An **expression** is something that composes a DSP graph rather than a
// value. Its two branches are the two def families, which are peers:
// `SynthExpr` (`./ugens/graph.ts`) for the UGen graph, and `FaustExpr` here
// for Faust — a `Signal` (the signal API) or a `Box` (the box algebra).
//
// The two branches carry no common class of their own, and that is the one
// place this file reads differently from its Python sibling. There, `Expr` is
// a marker class under `AbstractObject`, so a runtime `isinstance` can ask
// "is this an expression at all?"; here the question is a compile-time one and
// the answer is the `Expr` union in `./asdef.ts`, which is what the ambient
// verbs accept. What both languages agree on is that `Signal` and `Box` share
// a roof and do not compose with each other: this class is that roof, and it
// does a real job — the operator vocabulary is `AbstractObject`'s, the *table*
// each family maps a selector through is its own (the box schema has no
// `lsh`/`rsh`, and `mod` is Faust's `rem` for a signal and `fmod` for a box).

import { AbstractObject } from "../base/absobject.ts";
import type { Composed, Fan } from "../base/absobject.ts";

/** What a math method answers here: nothing fans out, so always the receiver. */
type FaustResult<TSelf, T> = Composed<TSelf, T, Fan<never, never>>;

/**
 * An expression of a **Faust graph**: a `./signals.ts` `Signal` or a
 * `./boxes.ts` `Box`. What `FaustDef` compiles.
 *
 * A subclass implements two hooks — `composeBinop`, which takes the operand
 * order as a flag, and `unop` — and gets the whole operator vocabulary and the
 * reversed-operand methods from here. Python writes `1 - sig` and gets
 * `__rsub__`; a language with no operator overloading needs that case spelled
 * out, which is what `rsub`/`rdiv` are.
 */
export abstract class FaustExpr<TSelf, TOperand>
    extends AbstractObject<TSelf, TOperand> {
    /**
     * Builds the node for `selector` over this expression and `other`, in
     * written order unless `swap` puts `other` first — the two Python hooks
     * (`_compose_binop`, `_rcompose_binop`) as one.
     */
    protected abstract composeBinop(
        selector: string,
        other: TOperand,
        swap: boolean,
    ): TSelf;

    protected binop<T extends TOperand>(
        selector: string,
        other: T,
    ): FaustResult<TSelf, T> {
        return this.composeBinop(selector, other, false) as FaustResult<TSelf, T>;
    }

    /** This expression on the **right** of a subtraction: `x - this`. */
    rsub(x: TOperand): TSelf { return this.composeBinop("sub", x, true); }
    /** `x / this`; see `rsub`. */
    rdiv(x: TOperand): TSelf { return this.composeBinop("div", x, true); }
}
